homeassistant/components/notify/matrix.py

```diff
@@ -0,0 +1,177 @@
+        config.get(CONF_USERNAME, None),
+        config.get(CONF_PASSWORD, None)
+    )
+
+
+# pylint: disable=too-few-public-methods
+class MatrixNotificationService(BaseNotificationService):
+    """Wrapper for the MatrixNotificationClient."""
+
+    # pylint: disable=too-many-arguments
+    def __init__(self, homeserver, default_room, session_file, verify_ssl,
+                 username=None, password=None):
+        """Buffer configuration data for send_message."""
+        self.homeserver = homeserver
+        self.default_room = default_room
+        self.session_file = session_file
+        self.verify_tls = verify_ssl
+        self.username = username
+        self.password = password
+
+    def send_message(self, message, **kwargs):
+        """Wrapper function to instantiate a MatrixNotificationClient."""
+        target_rooms = kwargs.get(ATTR_TARGET) or [self.default_room]
+        send_message(
+            message,
+            self.homeserver,
+            target_rooms,
+            self.session_file,
+            self.verify_tls,
+            self.username,
+            self.password
+        )
+
+
+# pylint: disable=too-many-arguments
+def send_message(message, homeserver, target_rooms, session_file,
+                 verify_tls=True, username=None, password=None):
+    """Do everything thats necessary to send a message to a Matrix room."""
+    from matrix_client.client import MatrixClient, MatrixRequestError
+
+    class MatrixNotificationClient:
+        """Handle Homeserver connection, authentication and messaging."""
+
+        def __init__(self):
+            """Set up a connection to the homeserver."""
+            self.client = None
+            self.token = None
+
+            self.session_file = session_file
+            self.load_token()
+            if self.token:
+                if not self.login_by_token():
+                    if not self.login_by_password():
+                        _LOGGER.error(
+                            'login failed, both token and username/password '
+                            'invalid'
+                        )
+                        return
+            else:
+                if not self.login_by_password():
+                    _LOGGER.error('login failed, username/password invalid')
+                    return
+
+            rooms = self.client.get_rooms()
+            for target_room in target_rooms:
+                try:
+                    if target_room in rooms:
+                        room = rooms[target_room]
+                    else:
+                        room = self.client.join_room(target_room)
+
+                    _LOGGER.debug(room.send_text(message))
+                except MatrixRequestError as ex:
+                    _LOGGER.error(
+                        'Unable to deliver message to room \'%s\': (%d): %s',
+                        target_room, ex.code, ex.content
+                    )
+
+        def load_token(self):
```

Comment:
Since load_token get called on each message sent, can't we cache it in a global var?

```python
_TOKENS = {}
...

if sessionfile in _TOKENS:
  return _TOKENS[sessionfile]
... continue with load
_TOKENS[sessionfile] = loaded
```